cmd/juicify/app/juicify.go

```diff
@@ -5,66 +5,124 @@ package app
+	AccessToken string   `json:"accessToken,omitempty"`
 
-	Servers []string `json:"servers"`
+	Requirements restapi.SessionRequirements `json:"requirements,omitempty"`
+}
 
-	LogGroup string `json:"logGroup,omitempty"`
+var (
+	controller  = flag.String("controller", "", "The IP address or hostname and port of the controller")
+	accessToken = flag.String("access-token", "", "The access token to use when connecting to the controller")
 
-	LogFile string `json:"logFile,omitempty"`
+	test           = flag.Bool("test", false, "Deprecated: Use --test-connection instead")
+	testConnection = flag.Bool("test-connection", false, "Tests the reachability of the controller or server(s)")
 
-	ForceSoftwareDecode bool `json:"forceSoftwareDecode,omitempty"`
+	juicePath = flag.String("juice-path", "", "Path to the juice executables if different than current executable path")
 
-	DisableCache bool `json:"disableCache,omitempty"`
+	errCancelled            = errors.New("cancelled")
+	errInvalidConfiguration = errors.New("invalid configuration")
+)
 
-	DisableCompression bool `json:"disableCompression,omitempty"`
+func validateConfiguration(config *Configuration) error {
+	if config.Id != "" {
+		return errInvalidConfiguration.Wrap(errors.New("id must not be specified"))
+	} else if len(config.Servers) != 0 {
+		return errInvalidConfiguration.Wrap(errors.New("servers must not be specified"))
+	} else if config.Controller == "" {
+		return errInvalidConfiguration.Wrap(errors.New("controller must be specified"))
+	} else if len(config.Requirements.Gpus) == 0 {
+		config.Requirements.Gpus = append(config.Requirements.Gpus, restapi.GpuRequirements{})
+	}
 
-	Headless bool `json:"headless,omitempty"`
+	return nil
+}
 
-	AllowTearing bool `json:"allowTearing,omitempty"`
+func requestSession(group task.Group, client *http.Client, config *Configuration) error {
+	api := restapi.Client{
+		Client:      client,
+		Address:     config.Controller,
+		AccessToken: config.AccessToken,
+	}
 
-	FrameRateLimit int `json:"frameRateLimit,omitempty"`
+	id, err := api.RequestSessionWithContext(group.Ctx(), config.Requirements)
+	if err != nil {
+		return err
+	}
 
-	FramesQueueAhead int `json:"framesQueueAhead,omitempty"`
+	session, err := api.GetSessionWithContext(group.Ctx(), id)
+	if err != nil {
+		return err
+	}
 
-	SwapChainBuffers int `json:"swapChainBuffers,omitempty"`
+	if session.State == restapi.SessionQueued {
+		logger.Info("Session queued")
 
-	WaitForDebugger bool `json:"waitForDebugger,omitempty"`
+		ticker := time.NewTicker(5 * time.Second)
+		defer ticker.Stop()
 
-	PCIBus []string `json:"pcibus,omitempty"`
+		for session.State == restapi.SessionQueued {
+			select {
+			case <-group.Ctx().Done():
+				err = errCancelled
 
-	AccessToken string `json:"accessToken,omitempty"`
+			case <-ticker.C:
+				session, err = api.GetSessionWithContext(group.Ctx(), id)
+			}
+
+			if err != nil {
+				logger.Info("Cancelling session")
+
+				api.CancelSession(id)
```

Comment:
can we also call this when the user ctrl-c's out of the session?

Reply:
Going to put this in another PR. There are a few places where it needs to go.